File: javae/src/IceInternal/IncomingBase.java
// **********************************************************************
//
// Copyright (c) 2003-2005 ZeroC, Inc. All rights reserved.
//
// This copy of Ice-E is licensed to you under the terms described in the
// ICE_LICENSE file included in this distribution.
//
// **********************************************************************

package IceInternal;

public class IncomingBase
{
    protected
    IncomingBase(Instance instance, Ice.Connection connection, Ice.ObjectAdapter adapter, boolean response)
    {
	_response = response;
        _os = new BasicStream(instance);
	_connection = connection;

        _current = new Ice.Current();
        _current.id = new Ice.Identity();
        _current.adapter = adapter;
        _current.con = _connection;

        _cookie = new Ice.LocalObjectHolder();
    }

    protected
    IncomingBase(IncomingBase in) // Adopts the argument. It must not be used afterwards.
    {
	_current = in._current;
	in._current = null;

	_servant = in._servant;
	in._servant = null;

	_cookie = in._cookie;
	in._cookie = null;

	_response = in._response;
	in._response = false;

	_os = in._os;
	in._os = null;

	_connection = in._connection;
	in._connection = null;
    }

    //
    // This function allows this object to be reused, rather than reallocated.
    //
    public void
    reset(Instance instance, Ice.Connection connection, Ice.ObjectAdapter adapter, boolean response)
    {
        //
        // Don't recycle the Current object, because servants may keep a reference to it.
        //
        _current = new Ice.Current();
        _current.id = new Ice.Identity();
        _current.adapter = adapter;
        _current.con = connection;

	_servant = null;

	if(_cookie == null)
	{
	    _cookie = new Ice.LocalObjectHolder();
	}
	else
	{
	    _cookie.value = null;
	}

	_response = response;

	if(_os == null)
	{
	    _os = new BasicStream(instance);
	}
	else
	{
	    _os.reset();
	}

	_connection = connection;
    }

    final protected void
    __warning(java.lang.Exception ex)
    {
	if(IceUtil.Debug.ASSERT)
	{
	    IceUtil.Debug.Assert(_os != null);
	}

	StringBuffer sb = new StringBuffer();
	sb.append("dispatch exception:");
	sb.append("\nidentity: " + Ice.Util.identityToString(_current.id));
	sb.append("\nfacet: " + IceUtil.StringUtil.escapeString(_current.facet, ""));
	sb.append("\noperation: " + _current.operation);
	sb.append("\n");
	sb.append(ex.toString());
	_os.instance().logger().warning(sb.toString());
    }

    protected Ice.Current _current;
    protected Ice.Object _servant;
    protected Ice.LocalObjectHolder _cookie;

    protected boolean _response;

    protected BasicStream _os;

    protected Ice.Connection _connection;
}

File: cs/src/Ice/IncomingBase.cs
// **********************************************************************
//
// Copyright (c) 2003 - 2004
// ZeroC, Inc.
// North Palm Beach, FL, USA
//
// All Rights Reserved.
//
// This copy of Ice is licensed to you under the terms described in the
// ICE_LICENSE file included in this distribution.
//
// **********************************************************************


namespace IceInternal
{

    using System.Diagnostics;

    public class IncomingBase
    {
	protected internal IncomingBase(Instance instance, Connection connection,
	                                Ice.ObjectAdapter adapter, bool response, byte compress)
	{
	    _current = new Ice.Current();
	    _current.id = new Ice.Identity();
	    _current.adapter = adapter;
	    
	    _cookie = null;
	    
	    _response = response;
	    
	    _compress = compress;
	    
	    _os = new BasicStream(instance);
	    
	    _connection = connection;
	}
	
	protected internal IncomingBase(IncomingBase inc)
	{
	    _current = inc._current;
	    inc._current = null;
	    
	    _servant = inc._servant;
	    inc._servant = null;
	    
	    _locator = inc._locator;
	    inc._locator = null;
	    
	    _cookie = inc._cookie;
	    inc._cookie = null;
	    
	    _response = inc._response;
	    inc._response = false;
	    
	    _compress = inc._compress;
	    inc._compress = 0;
	    
	    _os = inc._os;
	    inc._os = null;
	    
	    _connection = inc._connection;
	    inc._connection = null;
	}
	
	//
	// Do NOT use a finalizer, this would cause a severe performance
	// penalty! We must make sure that __destroy() is called instead,
	// to reclaim resources.
	//
	public virtual void __destroy()
	{
	    if(_os != null)
	    {
		_os.destroy();
		_os = null;
	    }
	}
	
	//
	// This function allows this object to be reused, rather than
	// reallocated.
	//
	public virtual void reset(Instance instance, Connection connection,
	                          Ice.ObjectAdapter adapter, bool response, byte compress)
	{
	    if(_current == null)
	    {
		_current = new Ice.Current();
		_current.id = new Ice.Identity();
		_current.adapter = adapter;
	    }
	    else
	    {
		Debug.Assert(_current.id != null);
		_current.adapter = adapter;
		
		if(_current.ctx != null)
		{
		    _current.ctx.Clear();
		}
	    }
	    
	    _servant = null;
	    
	    _locator = null;
	    
	    _cookie = null;
	    
	    _response = response;
	    
	    _compress = compress;
	    
	    if(_os == null)
	    {
		_os = new BasicStream(instance);
	    }
	    else
	    {
		_os.reset();
	    }
	    
	    _connection = connection;
	}
	
	protected internal void __finishInvoke()
	{
	    if(_locator != null && _servant != null)
	    {
		_locator.finished(_current, _servant, _cookie);
	    }
	    
	    //
	    // Send a response if necessary. If we don't need to send a
	    // response, we still need to tell the connection that we're
	    // finished with dispatching.
	    //
	    if(_response)
	    {
		_connection.sendResponse(_os, _compress);
	    }
	    else
	    {
		_connection.sendNoResponse();
	    }
	}
	
	protected internal void __warning(System.Exception ex)
	{
	    Debug.Assert(_os != null);
	    
	    IceUtil.OutputBase output = new IceUtil.OutputBase("");
	    output.setUseTab(false);
	    output.print("dispatch exception:");
	    output.print("\nidentity: " + Ice.Util.identityToString(_current.id));
	    output.print("\nfacet: ");
	    IceInternal.ValueWriter.write(_current.facet, output);
	    output.print("\noperation: " + _current.operation);
	    output.print("\n");
	    output.print(ex.ToString());
	    _os.instance().logger().warning(output.ToString()); // TODO: need to override ToString()?
	}
	
	protected internal Ice.Current _current;
	protected internal Ice.Object _servant;
	protected internal Ice.ServantLocator _locator;
	protected internal Ice.LocalObject _cookie;
	
	protected internal bool _response;
	protected internal byte _compress;
	
	protected internal BasicStream _os;
	
	private Connection _connection;
    }
	
}
